### PR TITLE
[contractstaking] fix invalid delta transfer

### LINKED ISSUE #3921: indexer lagging behind?
### What version of iotex-core image (or code branch) are you using?
v1.11.0

### What operating system and processor architecture are you using?
GCP testnet cluster

### What did you do? If possible, provide a recipe for reproducing the error.
node keeps generating this error
```
2023-08-14T19:43:55.571Z	ERROR	blockchain/pubsubmanager.go:121	Failed to handle new block.	{"ioAddr": "io1dg65erd07hrvyme0493f2kqj2utuvpyf6jeuhd", "error": "wrong block height 21768565, expecting 21693700: invalid input", "errorVerbose": "invalid input\ngithub.com/iotexproject/iotex-core/db.init\n\t/go/apps/iotex-core/db/counting_index.go:24\nruntime.doInit\n\t/usr/local/go/src/runtime/proc.go:6222\nruntime.doInit\n\t/usr/local/go/src/runtime/proc.go:6199\nruntime.doInit\n\t/usr/local/go/src/runtime/proc.go:6199\nruntime.doInit\n\t/usr/local/go/src/runtime/proc.go:6199\nruntime.doInit\n\t/usr/local/go/src/runtime/proc.go:6199\nruntime.main\n\t/usr/local/go/src/runtime/proc.go:233\nruntime.goexit\n\t/usr/local/go/src/runtime/asm_amd64.s:1571\nwrong block height 21768565, expecting 21693700\ngithub.com/iotexproject/iotex-core/blockindex.(*blockIndexer).putBlock\n\t/go/apps/iotex-core/blockindex/indexer.go:305\ngithub.com/iotexproject/iotex-core/blockindex.(*blockIndexer).PutBlock\n\t/go/apps/iotex-core/blockindex/indexer.go:140\ngithub.com/iotexproject/iotex-core/blockindex.(*IndexBuilder).ReceiveBlock\n\t/go/apps/iotex-core/blockindex/indexbuilder.go:99\ngithub.com/iotexproject/iotex-core/blockchain.(*pubSub).handler\n\t/go/apps/iotex-core/blockchain/pubsubmanager.go:120\nruntime.goexit\n\t/usr/local/go/src/runtime/asm_amd64.s:1571"}
```

### What did you expect to see?
this should not happen, since indexer should commit every new block and update height

### What did you see instead?
seems the indexer is lagging behind

### PR DESCRIPTION
**Pull request by envestcc**
# Description

`Stake` and `Merge` in one transaction will cause action `deltaActionRemove` on state `deltaStateAdded`, which is not allowed currently. But this scenario should be legal, so this transfer should be allowed instead. 

Fixes #3921 

## Type of change
Please delete options that are not relevant.
- [] Bug fix (non-breaking change which fixes an issue)
- [] New feature (non-breaking change which adds functionality)
- [] Code refactor or improvement
- [x] Breaking change (fix or feature that would cause a new or changed behavior of existing functionality)
- [] This change requires a documentation update

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration
- [x] make test
- [] fullsync
- [] Other test (please specify)

**Test Configuration**:
- Firmware version:
- Hardware:
- Toolchain:
- SDK:

# Checklist:
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: blockindex/contractstaking/delta_state.go [hunk 1]
```diff
@@ -27,6 +27,7 @@ var (
 		},
 		deltaStateAdded: {
 			deltaActionModify: deltaStateAdded,
+			deltaActionRemove: deltaStateUnchanged,
```
Comment (dustinxie) [inline blockindex/contractstaking/delta_state.go:30]: what about `deltaStateRemoved`? does it need a state-transition?

Comment (dustinxie) [reply blockindex/contractstaking/delta_state.go:30]: also need to check `blkHeight == currentHeight + 1` in `PutBlock()`?

Comment (envestcc) [reply blockindex/contractstaking/delta_state.go:30]: > what about `deltaStateRemoved`? does it need a state-transition?

no valid action could be triggered on `deltaStateRemoved`

Comment (envestcc) [reply blockindex/contractstaking/delta_state.go:30]: > as we chatted, also need to check `blkHeight == currentHeight + 1` in `PutBlock()`?

Thx for reminding

File: blockindex/contractstaking/indexer.go [hunk 2]
```diff
@@ -121,9 +121,16 @@ func (s *Indexer) BucketTypes() ([]*BucketType, error) {
 
 // PutBlock puts a block into indexer
 func (s *Indexer) PutBlock(ctx context.Context, blk *block.Block) error {
-	if blk.Height() < s.contractDeployHeight || blk.Height() <= s.cache.Height() {
+	expectHeight := s.cache.Height() + 1
+	if expectHeight < s.contractDeployHeight {
+		expectHeight = s.contractDeployHeight
+	}
+	if blk.Height() < expectHeight {
```
Comment (dustinxie) [inline blockindex/contractstaking/indexer.go:128]: let's still use
```
if blk.Height() < s.contractDeployHeight {
    return nil
}
expectHeight := s.cache.Height() + 1
...
```
i think this is more clear and easy to understand

Comment (envestcc) [reply blockindex/contractstaking/indexer.go:128]: `expectHeight` will be 1 when `blk.Height() == s.contractDeployHeight` if in this way

File: blockindex/contractstaking/indexer.go [hunk 3]
```diff
@@ -121,9 +121,16 @@ func (s *Indexer) BucketTypes() ([]*BucketType, error) {
 
 // PutBlock puts a block into indexer
 func (s *Indexer) PutBlock(ctx context.Context, blk *block.Block) error {
-	if blk.Height() < s.contractDeployHeight || blk.Height() <= s.cache.Height() {
+	expectHeight := s.cache.Height() + 1
```
Comment (dustinxie) [inline blockindex/contractstaking/indexer.go:124]: remove empty line

File: blockindex/contractstaking/delta_state.go (hunk 1 above)
Comment (dustinxie) [reply blockindex/contractstaking/delta_state.go:30]: add a test to validate, once a bucket is deleted, i.e., `deltaStateRemoved`, it cannot transition to any further state (in the same block)

**Review by millken — APPROVED** [review]

**Pull request merged**